tests/unit/ip_router_interface/v0/test_ip_router.py

```diff
@@ -0,0 +1,487 @@
+# Copyright 2023 Ubuntu
+# See LICENSE file for licensing details.
+#
+# Learn more about testing at: https://juju.is/docs/sdk/testing
+import json
+import logging
+import os
+import shutil
+import unittest
+from itertools import chain
+
+from ops.testing import Harness
+
+from tests.provider_charm.src.charm import (
+    IP_ROUTER_PROVIDER_RELATION_NAME,
+    SimpleIPRouteProviderCharm,
+)
+from tests.requirer_charm.src.charm import (
+    IP_ROUTER_REQUIRER_RELATION_NAME,
+    SimpleIPRouteRequirerCharm,
+)
+
+logger = logging.getLogger(__name__)
+
+LIB_DIR = "lib/charms/ip_router_interface/v0"
+LIB_NAME = "ip_router_interface.py"
+REQUIRER_CHARM_DIR = "tests/requirer_charm"
+PROVIDER_CHARM_DIR = "tests/provider_charm"
+IP_ROUTER_PROVIDER_APP_NAME = "ip-router-provider"
+IP_ROUTER_REQUIRER_APP_NAME = "ip-router-requirer"
+
+
+def copy_lib_content() -> None:
+    os.makedirs(f"{REQUIRER_CHARM_DIR}/{LIB_DIR}", exist_ok=True)
+    os.makedirs(f"{PROVIDER_CHARM_DIR}/{LIB_DIR}", exist_ok=True)
+    shutil.copyfile(src=f"{LIB_DIR}/{LIB_NAME}", dst=f"{REQUIRER_CHARM_DIR}/{LIB_DIR}/{LIB_NAME}")
+    shutil.copyfile(src=f"{LIB_DIR}/{LIB_NAME}", dst=f"{PROVIDER_CHARM_DIR}/{LIB_DIR}/{LIB_NAME}")
+
+
+class TestProvider(unittest.TestCase):
+    def setUp(self):
+        copy_lib_content()
+        harness = Harness(SimpleIPRouteProviderCharm)
+        harness.set_model_name("test")
+        self.addCleanup(harness.cleanup)
+        harness.begin()
+        harness.set_leader()
+        self.harness = harness
+
+    def test_provider_initial_setup(self):
+        # Check initial routing table
+        assert self.harness.charm.RouterProvider.get_routing_table() == {}
+
+    def test_given_provider_when_network_request_received_then_adds_network(self):
+        # Create a relation
+        rel_id = self.harness.add_relation(IP_ROUTER_PROVIDER_RELATION_NAME, "ip-router-requirer")
+        self.harness.add_relation_unit(rel_id, "ip-router-requirer/0")
+
+        # Update databag
+        network_request = [{"network": "192.168.250.0/24", "gateway": "192.168.250.1"}]
+        self.harness.update_relation_data(
+            rel_id, "ip-router-requirer", {"networks": json.dumps(network_request)}
+        )
+        self.harness.update_relation_data(
+            rel_id, "ip-router-requirer", {"network-name": IP_ROUTER_REQUIRER_RELATION_NAME}
+        )
+
+        # Get routing table
+        expected_rt = {IP_ROUTER_REQUIRER_RELATION_NAME: network_request}
+        assert self.harness.charm.RouterProvider.get_routing_table() == expected_rt
+        expected_databag = {"networks": json.dumps(network_request)}
+        assert (
+            self.harness.get_relation_data(rel_id, self.harness.charm.app.name) == expected_databag
+        )
+
+    def test_given_provider_when_network_request_with_provider_received_then_adds_network(self):
+        # Create a relation
```

Comment:
A lot of the comments feel unnecessary and inconsistent. For instance why are we saying that we are creating a relation here but not that we are adding a unit to the relation. Overall I'd suggest getting rid of all of them.

Reply:
Removed